### PR TITLE
new resource and data source: azurerm_oracle_exadata_infrastructure

### PR DESCRIPTION
**Pull request by eelhomsi**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Introduce OracleDatabase ExaData Infrastructure Resource & Data
Most relevant changes in:
`internal/services/oracledatabase/`

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

`ExadataInfraDataSource` test:
Screenshot 2024-08-09 at 11 18 53

`ExadataInfraResource` tests:
Screenshot 2024-08-09 at 11 10 32

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* FEATURES - Introduced new ExadataInfraDataSource and ExadataInfraResource

This is a (please select all that apply):

- [ ] Bug Fix
- [X] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: website/docs/r/oracledatabase_exadata_infrastructure.html.markdown [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  compute_count = 2
+  shape = "Exadata.X9M"
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `compute_count` - (Required) The number of compute servers for the cloud Exadata infrastructure.
```
Comment (katbyte) [inline website/docs/r/oracledatabase_exadata_infrastructure.html.markdown:37]: might i suggest
```suggestion
* `compute_server_count` - (Required) The number of compute servers for the cloud Exadata infrastructure.
```

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+	MaintenanceWindow []MaintenanceWindowModel `tfschema:"maintenance_window"`
+}
+
+func (ExadataInfraResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		// Azure
+		"location": commonschema.Location(),
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:49]: please validate everything possible by adding validation functions

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		},
+		"resource_group_name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+		"tags":  commonschema.Tags(),
+		"zones": commonschema.ZonesMultipleRequired(),
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:56]: these should be at the bottom
```suggestion
		"tags":  commonschema.Tags(),
		
		"zones": commonschema.ZonesMultipleRequired(),
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/oracledatabase_exadata_infrastructure.html.markdown (hunk 1 above)
Comment (eelhomsi) [reply website/docs/r/oracledatabase_exadata_infrastructure.html.markdown:37]: I agree that this is clearer - however we have a requirement to try and ensure the argument names are the same as those in the official OCI Terraform provider:
https://registry.terraform.io/providers/oracle/oci/latest/docs/resources/database_cloud_exadata_infrastructure#compute_count

File: internal/services/oracledatabase/exadata_infrastructure_resource.go (hunk 3 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_resource.go:56]: Done thanks!

File: internal/services/oracledatabase/exadata_infrastructure_resource.go (hunk 2 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_resource.go:49]: Done, thanks!

File: website/docs/r/oracledatabase_exadata_infrastructure.html.markdown [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+* `days_of_week` - (Optional) Days during the week when maintenance should be performed. Valid values are: 0 - represents time slot 0:00 - 3:59 UTC - 4 - represents time slot 4:00 - 7:59 UTC - 8 - represents time slot 8:00 - 11:59 UTC - 12 - represents time slot 12:00 - 15:59 UTC - 16 - represents time slot 16:00 - 19:59 UTC - 20 - represents time slot 20:00 - 23:59 UTC
+
+* `hours_of_day` - (Optional) The window of hours during the day when maintenance should be performed. The window is a 4 hour slot.
+
+* `is_custom_action_timeout_enabled` - (Optional) If true, enables the configuration of a custom action timeout (waiting period) between database server patching operations.
+
+* `is_monthly_patching_enabled` - (Optional) If true, enables the monthly patching option.
+
+* `lead_time_in_weeks` - (Optional) Lead time window allows user to set a lead time to prepare for a down time. The lead time is in weeks and valid value is between 1 to 4.
```
Comment (katbyte) [inline website/docs/r/oracledatabase_exadata_infrastructure.html.markdown:75]:
```suggestion
* `lead_time_in_weeks` - (Optional) Lead time window allows user to set a lead time to prepare for a down time. The lead time is in weeks and valid value is between `1` to `4`.
```
values should be quoted as such

File: website/docs/r/oracledatabase_exadata_infrastructure.html.markdown [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Exadata Infrastructure.
+
+---
+
+A `maintenance_window` block supports the following:
+
+* `custom_action_timeout_in_mins` - (Optional) Determines the amount of time the system will wait before the start of each database server patching operation. Custom action timeout is in minutes and valid value is between 15 to 120 (inclusive).
+
+* `days_of_week` - (Optional) Days during the week when maintenance should be performed. Valid values are: 0 - represents time slot 0:00 - 3:59 UTC - 4 - represents time slot 4:00 - 7:59 UTC - 8 - represents time slot 8:00 - 11:59 UTC - 12 - represents time slot 12:00 - 15:59 UTC - 16 - represents time slot 16:00 - 19:59 UTC - 20 - represents time slot 20:00 - 23:59 UTC
```
Comment (katbyte) [inline website/docs/r/oracledatabase_exadata_infrastructure.html.markdown:67]: > 
```suggestion
* `days_of_week` - (Optional) Days during the week when maintenance should be performed. Valid values are: `0` - represents time slot `0:00 - 3:59 UTC - 4` - represents time slot `4:00 - 7:59 UTC - 8` - represents time slot 8:00 - 11:59 UTC - 12 - represents time slot 12:00 - 15:59 UTC - 16 - represents time slot 16:00 - 19:59 UTC - 20 - represents time slot `20:00 - 23:59 UTC`
```

etc

File: website/docs/r/oracledatabase_exadata_infrastructure.html.markdown [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+* `maintenance_window` - (Optional) One or more `maintenance_window` blocks as defined below.
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Exadata Infrastructure.
+
+---
+
+A `maintenance_window` block supports the following:
+
+* `custom_action_timeout_in_mins` - (Optional) Determines the amount of time the system will wait before the start of each database server patching operation. Custom action timeout is in minutes and valid value is between 15 to 120 (inclusive).
```
Comment (katbyte) [inline website/docs/r/oracledatabase_exadata_infrastructure.html.markdown:65]:
```suggestion
* `custom_action_timeout_in_mins` - (Optional) Determines the amount of time the system will wait before the start of each database server patching operation. Custom action timeout is in minutes and valid value is between `15` to `120` (inclusive).
```

File: internal/services/oracledatabase/registration.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+// Name is the name of this Service
+func (r Registration) Name() string {
+	return "App Service"
+}
+
+// WebsiteCategories returns a list of categories which can be used for the sidebar
+func (r Registration) WebsiteCategories() []string {
+	return []string{
+		"App Service",
```
Comment (katbyte) [inline internal/services/oracledatabase/registration.go:35]: this is wrong?

File: internal/services/oracledatabase/registration.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+func (r Registration) Resources() []sdk.Resource {
+	return []sdk.Resource{
+		ExadataInfraResource{},
+	}
+}
+
+// Name is the name of this Service
+func (r Registration) Name() string {
+	return "App Service"
```
Comment (katbyte) [inline internal/services/oracledatabase/registration.go:29]: and here?

File: internal/services/oracledatabase/exadata_infrastructure_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    hours_of_day                     = [1]
+    is_custom_action_timeout_enabled = false
+    is_monthly_patching_enabled      = true
+    months                           = ["January"]
+    weeks_of_month                   = [2]
+    lead_time_in_weeks               = 1
+    patching_mode                    = "Rolling"
+    preference                       = "NoPreference"
+  }
+  tags = {
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource_test.go:155]: we should make this tidier
```suggestion
  customer_contacts   = ["test@test.com"]

  maintenance_window {
    custom_action_timeout_in_mins    = 0
    days_of_week                     = ["Monday"]
    hours_of_day                     = [1]
    is_custom_action_timeout_enabled = false
    is_monthly_patching_enabled      = true
    months                           = ["January"]
    weeks_of_month                   = [2]
    lead_time_in_weeks               = 1
    patching_mode                    = "Rolling"
    preference                       = "NoPreference"
  }

  tags = {
```

File: internal/services/oracledatabase/exadata_infrastructure_resource_test.go [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+		},
+		data.ImportStep(),
+	})
+}
+
+func (a ExadataInfraResource) basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+
+
+%s
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource_test.go:112]: can we tidy this up??

```suggestion
	return fmt.Sprintf(`
%s
```

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+			output.Location = result.Model.Location
+			output.Zones = result.Model.Zones
+			output.ResourceGroupName = id.ResourceGroupName
+			output.Tags = utils.FlattenPtrMapStringString(result.Model.Tags)
+			prop := result.Model.Properties
+			output.ComputeCount = pointer.From(prop.ComputeCount)
+			output.DisplayName = prop.DisplayName
+			output.StorageCount = pointer.From(prop.StorageCount)
+			output.Shape = prop.Shape
+			output.MaintenanceWindow = FlattenMaintenanceWindow(prop.MaintenanceWindow)
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:328]: minor issue but to match other resources we should inline this/do an assingment

```
output := ExadataInfraResourceModel {
CustomerContacts: FlattenCustomerContacts(result.Model.Properties.CustomerContacts)
Name: pointer.ToString(result.Model.Name)
...
```
and such vs going outout.x

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+						Optional: true,
+						Computed: true,
+					},
+
+					"weeks_of_month": {
+						Type:     pluginsdk.TypeList,
+						Optional: true,
+						Computed: true,
+						Elem: &pluginsdk.Schema{
+							Type: pluginsdk.TypeInt,
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:166]: this should be validated between 1-5?

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+					"patching_mode": {
+						Type:     pluginsdk.TypeString,
+						Optional: true,
+						Computed: true,
+					},
+
+					"preference": {
+						Type:     pluginsdk.TypeString,
+						Optional: true,
+						Computed: true,
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:158]: this should be validated?

File: internal/services/oracledatabase/exadata_infrastructure_resource.go [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+						Computed:     true,
+						ValidateFunc: validate.LeadTimeInWeeks,
+					},
+
+					"months": {
+						Type:     pluginsdk.TypeList,
+						Optional: true,
+						Computed: true,
+						Elem: &pluginsdk.Schema{
+							Type: pluginsdk.TypeString,
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_resource.go:145]: and here? 

please validate all properties

File: internal/services/oracledatabase/exadata_infrastructure_data_source.go [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+							LastModifiedAt:     systemData.LastModifiedAt,
+						},
+					}
+				}
+				output.Name = id.CloudExadataInfrastructureName
+				output.ResourceGroupName = id.ResourceGroupName
+				output.Type = pointer.From(model.Type)
+				output.Tags = utils.FlattenPtrMapStringString(model.Tags)
+				output.Location = model.Location
+				output.Zones = model.Zones
```
Comment (katbyte) [inline internal/services/oracledatabase/exadata_infrastructure_data_source.go:502]: why are these seperate and not included above?

**Review by katbyte — CHANGES_REQUESTED** [review]

File: internal/services/oracledatabase/registration.go (hunk 7 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/registration.go]: Fixed, thanks!

File: internal/services/oracledatabase/exadata_infrastructure_resource.go (hunk 11 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_resource.go]: Understood, I've updated the code.

File: internal/services/oracledatabase/exadata_infrastructure_resource.go (hunk 12 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_resource.go]: Internally we want 1-4, I've updated to validate that, thank you.

File: internal/services/oracledatabase/exadata_infrastructure_resource.go (hunk 14 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_resource.go]: Done!

File: internal/services/oracledatabase/exadata_infrastructure_data_source.go (hunk 15 above)
Comment (eelhomsi) [reply internal/services/oracledatabase/exadata_infrastructure_data_source.go]: I've moved them up with the test.

**Review by katbyte — APPROVED** [review]
I think this now looks good provided CI passes, please show confirmation of passing tests and we can get this merged

**Comment by eelhomsi** [comment]
> I think this now looks good provided CI passes, please show confirmation of passing tests and we can get this merged

I've added testing evidence here:
Screenshot 2024-10-04 at 01 08 57
Screenshot 2024-10-04 at 01 01 33

**Comment by katbyte** [comment]
@eelhomsi please run `make terrafmt` to fix ci

**Comment by eelhomsi** [comment]
> @eelhomsi please run `make terrafmt` to fix ci

Done

**Comment by eelhomsi** [comment]
Latest tests here:
Screenshot 2024-10-11 at 20 10 59
Screenshot 2024-10-11 at 19 55 58

**Comment by jackofallops** [comment]
Hi @eelhomsi - Apologies again, but we need to revert this one also. Can you please re-open the PR from your branch and we'll work with you to get this in asap.

Thanks

**Comment by eelhomsi** [comment]
> Hi @eelhomsi - Apologies again, but we need to revert this one also. Can you please re-open the PR from your branch and we'll work with you to get this in asap.
> 
> Thanks

No problem @jackofallops PR reopened here:
https://github.com/hashicorp/terraform-provider-azurerm/pull/27678

**Pull request merged**